### DISCUSSION TITLE
request: graph query (bi-directions; forward and reverse) for label

### DISCUSSION THREAD

**Discussion by itaru2622** [bluesky-social/atproto#2659] [General Discussion]
current API supports forward query but none for reverse query, described in below figure.
it should support graph query (bi-directional queries) for label, similar to graph query for follows/followers.

![image](https://github.com/user-attachments/assets/fd6d58f1-0e07-458b-bfc9-2f91718137a0)

**Comment by verdverm** [comment]
Do you mean...?

currently exists: actor/post -> labels  (you can get the labels for a subject id)

desired (reverse): label -> actors/posts  (you want to be able to get all subjects that have a given id)

**Reply by itaru2622** [reply]
@verdverm I'm not sure what you cannot understand.

What I wanted to say is:
relationship between labeler and target of label(actor or post) is similar to the relationship among users(follows and followers).
i.e: each labeler adds label(s) to target(s), then one target may got labels from labelers.

so it is desired to support graph query (bi-direction query) for label, similar to query for follows/followers.

**Reply by verdverm** [reply]
Maybe show some examples with data and the query response.

as I understand it
- you can ask a labeler for the labels they provide, their spec
- you can ask a labeler for the labels of a target

The gap I see is, you cannot ask a labeler for all the targets of a label, is that what you want?

**Comment by bnewbold** [comment]
Adding more query functionality for labels would make sense for some projects! I think we want to keep the "required" endpoints small and simple though. Could you give some more context and examples for what you are trying to do, and other projects that might also want to do the same queries? Could you achieve the same goal by running an independent service which aggregates labels?

**Reply by itaru2622** [reply]
By subscribing a specific labeler, user can obtain additional information that the labeler has assigned to the target(post or actor), such as AI generated image etc, and decide his/her action in some cases.

but without subscribing, user cannot know if the target is assigned label. moreover the number of labelers are already exceeded the upper limit of subscription(20) and still increasing day by day.

so, it is good when a reverse query gets all labels/labelers which assigned to specific target(post or actor) at once, like follows/followers graph query.

> Could you achieve the same goal by running an independent service which aggregates labels?

to achive by independent service with current APIs:
it requires forward queries(queryLabels) to all labelers(70+ and keep increasing), so it costs many queries on network in short period.
